gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/store/hdfs/SimpleHDFSConfigStore.java

```diff
@@ -0,0 +1,341 @@
+            configKeyPaths.addAll(Lists.newArrayList(Iterables
+                .transform(IOUtils.readLines(includesConfInStream, Charsets.UTF_8), new IncludesToConfigKey())));
+          }
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(String.format("Error while getting imports for configKey: \"%s\"", configKey), e);
+    }
+    return configKeyPaths;
+  }
+
+  /**
+   * Retrieves the {@link Config} for the given {@link ConfigKeyPath} by reading the {@code main.conf} associated with
+   * the dataset specified by the given {@link ConfigKeyPath}.
+   *
+   * @param  configKey      the config key path whose properties are needed.
+   * @param  version        the configuration version in the configuration store.
+   *
+   * @return a {@link Config} for the given configKey.
+   *
+   * @throws VersionDoesNotExistException if the version specified cannot be found in the {@link ConfigStore}.
+   */
+  @Override
+  public Config getOwnConfig(ConfigKeyPath configKey, String version)
+      throws VersionDoesNotExistException {
+    Preconditions.checkNotNull(configKey, "configKey cannot be null!");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(version), "version cannot be null or empty!");
+
+    Path configDir = getDatasetDirForKey(configKey, version);
+    checkDatasetDirExists(configKey, configDir);
+
+    try {
+      for (FileStatus fileStatus : this.fs.listStatus(configDir, new MainConfFileFilter())) {
+        if (!fileStatus.isDir()) {
+          try (InputStream mainConfInputStream = this.fs.open(fileStatus.getPath())) {
+            return ConfigFactory.parseReader(new InputStreamReader(mainConfInputStream)).resolve();
+          }
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(String.format("Error while getting config for configKey: \"%s\"", configKey), e);
+    }
+    return ConfigFactory.empty();
+  }
+
+  /**
+   * Retrieves the dataset dir on HDFS associated with the given {@link ConfigKeyPath} and the given version. This
+   * directory contains the {@code main.conf} and {@code includes.conf} file, as well as any child datasets.
+   */
+  private Path getDatasetDirForKey(ConfigKeyPath configKey, String version) throws VersionDoesNotExistException {
+    return new Path(getVersionRoot(version), getDatasetFromConfigKey(configKey));
+  }
+
+  /**
+   * Retrieves the name of a dataset from a given {@link ConfigKeyPath}.
+   */
+  private Path getDatasetFromConfigKey(ConfigKeyPath configKey) {
+    Preconditions.checkArgument(configKey.getAbsolutePathString().startsWith(this.physicalStoreRoot.getPath()),
+        String.format("The configKey \"%s\" must begin with \"%s\"!", configKey, this.physicalStoreRoot.getPath()));
+
+    return PathUtils.withoutLeadingSeparator(
+        new Path(StringUtils.removeStart(configKey.getAbsolutePathString(), this.physicalStoreRoot.getPath())));
```

Comment:
Basically it ensures the dataset doesn't being with `/` so that it can be combined with the result of `getVersionRoot`. I have updated the code a bit to make it clearer.